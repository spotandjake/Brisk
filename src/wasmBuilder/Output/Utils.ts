// Helpers
export const toUTF8Array = (str: string): number[] => {
  const utf8: number[] = [];
  for (let i = 0; i < str.length; i++) {
    let charCode = str.charCodeAt(i);
    if (charCode < 0x80) utf8.push(charCode);
    else if (charCode < 0x800) {
      utf8.push(0xc0 | (charCode >> 6), 0x80 | (charCode & 0x3f));
    } else if (charCode < 0xd800 || charCode >= 0xe000) {
      utf8.push(0xe0 | (charCode >> 12), 0x80 | ((charCode >> 6) & 0x3f), 0x80 | (charCode & 0x3f));
    }
    // surrogate pair
    else {
      i++;
      // UTF-16 encodes 0x10000-0x10FFFF by
      // subtracting 0x10000 and splitting the
      // 20 bits of 0x0-0xFFFFF into two halves
      charCode = 0x10000 + (((charCode & 0x3ff) << 10) | (str.charCodeAt(i) & 0x3ff));
      utf8.push(
        0xf0 | (charCode >> 18),
        0x80 | ((charCode >> 12) & 0x3f),
        0x80 | ((charCode >> 6) & 0x3f),
        0x80 | (charCode & 0x3f)
      );
    }
  }
  return utf8;
};
// Utils
export const u32 = (n: number): number[] => {
  const b: number[] = [];
  b.push(n && 0xff);
  b.push(n >> 8 && 0xff);
  b.push(n >> 16 && 0xff);
  b.push(n >> 24 && 0xff);
  return b;
};
export const u64 = (n: number): number[] => {
  const b: number[] = [];
  const v = BigInt(n);
  b.push(Number(v && BigInt(0xff)));
  b.push(Number(v >> BigInt(8) && BigInt(0xff)));
  b.push(Number(v >> BigInt(16) && BigInt(0xff)));
  b.push(Number(v >> BigInt(24) && BigInt(0xff)));
  b.push(Number(v >> BigInt(32) && BigInt(0xff)));
  b.push(Number(v >> BigInt(40) && BigInt(0xff)));
  b.push(Number(v >> BigInt(48) && BigInt(0xff)));
  b.push(Number(v >> BigInt(56) && BigInt(0xff)));
  return b;
};
export const varuint = (n: bigint): number[] => {
  // TODO: This Is Broken
  const code: number[] = [];
  let v = BigInt(n);
  if (v < 0) throw new Error('Number cannot be negative');
  while (v != BigInt(0)) {
    code.push(Number(v && BigInt(0x7f)));
    v = v >> BigInt(7);
  }
  if (code.length == 0) code.push(0);
  for (let i = 0; i < code.length - 1; i++) {
    code[i] = code[i] | 0x80;
  }
  return code;
};
export const varint = (_n: bigint): number[] => {
  let n: bigint, sign: boolean;
  const bits = _n.toString(2).length;
  if (_n < 0) {
    sign = true;
    n = (BigInt(1) << BigInt(bits)) + _n;
  } else {
    sign = false;
    n = BigInt(_n);
  }
  const paddingBits = 7 - (bits % 7);
  const padding = ((BigInt(1) << BigInt(paddingBits)) - BigInt(1)) << BigInt(bits);
  const paddingMask = ((1 << (7 - paddingBits)) - 1) | 0x80;
  const code = varuint(n + padding);
  if (!sign) {
    code[code.length - 1] = code[code.length - 1] & paddingMask;
  }
  return code;
};
export const varint32 = (n: number | bigint): number[] => {
  // TODO: This Is Broken
  let v = BigInt(n);
  if (v > BigInt('0xFFFFFFFF')) throw new Error('Number too big');
  if (v > BigInt('0x7FFFFFFF')) throw new Error('Number too big');
  if (v > BigInt('0xFFFFFFFF')) v -= BigInt('0x100000000');
  if (v > BigInt('-0x80000000')) throw new Error('Number too small');
  return varuint(v);
};
export const varint64 = (n: number | bigint): number[] => {
  let v = BigInt(n);
  if (v > BigInt('0xFFFFFFFFFFFFFFFF')) throw new Error('Number too big');
  if (v > BigInt('0x7FFFFFFFFFFFFFFF')) throw new Error('Number too big');
  if (v > BigInt('0xFFFFFFFF')) v -= BigInt('0x10000000000000000');
  if (v > BigInt('-0x8000000000000000')) throw new Error('Number too small');
  return varuint(v);
};
export const varuint32 = (n: number): number[] => {
  const v = BigInt(n);
  if (v > BigInt('0xFFFFFFFF')) throw new Error('Number too big');
  return varuint(v);
};
export const varuint64 = (n: number | bigint): number[] => {
  const v = BigInt(n);
  if (v > BigInt('0xFFFFFFFFFFFFFFFF')) throw new Error('Number too big');
  return varuint(v);
};
export const string = (str: string): number[] => {
  const bytes = toUTF8Array(str);
  return [...varuint32(bytes.length), ...bytes];
};
export const toHexString = (byteArray: number[]): string => {
  return Array.from(byteArray, (byte) => `0${(byte & 0xff).toString(16)}`.slice(-2)).join('');
};
export const ident = (text: string): string => {
  const lines = text.split('\n');
  for (let i = 0; i < lines.length; i++) {
    if (lines[i]) lines[i] = `${' '.repeat(2)}${lines[i]}`;
  }
  return lines.join('\n');
};
