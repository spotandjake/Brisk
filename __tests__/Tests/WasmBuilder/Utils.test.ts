// Test Utils
import { expect, test } from '@jest/globals';
// Test Components
import * as Utils from '../../../src/wasmBuilder/Build/Utils';
// WasmBuilder Utils Tests
test('WasmBuilder-Utils: ieee754-negative-ints', () => {
  const data = [];
  for (let i = 0; i >= -1000; i -= 250) {
    data.push(i);
  }
  expect(data.map((n) => Utils.ieee754(n).join(','))).toMatchSnapshot();
});
test('WasmBuilder-Utils: ieee754-positive-ints', () => {
  const data = [];
  for (let i = 0; i <= 1000; i += 250) {
    data.push(i);
  }
  expect(data.map((n) => Utils.ieee754(n).join(','))).toMatchSnapshot();
});
test('WasmBuilder-Utils: ieee754-negative-float', () => {
  const data = [];
  for (let i = 0; i >= -1; i -= 0.25) {
    data.push(i);
  }
  expect(data.map((n) => Utils.ieee754(n).join(','))).toMatchSnapshot();
});
test('WasmBuilder-Utils: ieee754-positive-float', () => {
  const data = [];
  for (let i = 0; i <= 1; i += 0.25) {
    data.push(i);
  }
  expect(data.map((n) => Utils.ieee754(n).join(','))).toMatchSnapshot();
});
test('WasmBuilder-Utils: ieee754-zero', () => {
  expect(Utils.ieee754(0)).toEqual(new Uint8Array([0x00, 0x00, 0x00, 0x00]));
});
test('WasmBuilder-Utils: signedLEB128-negative-ints', () => {
  const data = [];
  for (let i = 0; i >= -1000; i -= 250) {
    data.push(i);
  }
  expect(data.map((n) => Utils.signedLEB128(n).join(','))).toMatchSnapshot();
});
test('WasmBuilder-Utils: signedLEB128-positive-ints', () => {
  const data = [];
  for (let i = 0; i <= 1000; i += 100) {
    data.push(i);
  }
  expect(data.map((n) => Utils.signedLEB128(n).join(','))).toMatchSnapshot();
});
test('WasmBuilder-Utils: signedLEB128-zero', () => {
  expect(Utils.signedLEB128(0)).toEqual([0x00]);
});
test('WasmBuilder-Utils: unsignedLEB128-positive-ints', () => {
  const data = [];
  for (let i = 0; i <= 1000; i += 100) {
    data.push(i);
  }
  expect(data.map((n) => Utils.unsignedLEB128(n).join(','))).toMatchSnapshot();
});
test('WasmBuilder-Utils: unsignedLEB128-zero', () => {
  expect(Utils.unsignedLEB128(0).join(',')).toMatchSnapshot();
});
test('WasmBuilder-Utils: encodeVector', () => {
  const data = [
    ...[0x00, 0x61, 0x73, 0x6d], // Magic Module Header
    ...[0x01, 0x00, 0x00, 0x00], // Wasm Module Version
  ];
  expect(Utils.encodeVector(data)).toEqual([...Utils.unsignedLEB128(data.length), ...data]);
});
test('WasmBuilder-Utils: _encodeString', () => {
  expect(Utils._encodeString('Hello World').join(',')).toMatchSnapshot();
});
test('WasmBuilder-Utils: encodeString', () => {
  expect(Utils.encodeString('Hello World').join(',')).toMatchSnapshot();
});
