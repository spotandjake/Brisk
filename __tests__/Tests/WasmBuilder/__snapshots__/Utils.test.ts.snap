// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WasmBuilder-Utils: _encodeString 1`] = `"72,101,108,108,111,32,87,111,114,108,100"`;

exports[`WasmBuilder-Utils: encodeString 1`] = `"11,72,101,108,108,111,32,87,111,114,108,100"`;

exports[`WasmBuilder-Utils: ieee754-negative-float 1`] = `
Array [
  "0,0,0,0",
  "0,0,128,190",
  "0,0,0,191",
  "0,0,64,191",
  "0,0,128,191",
]
`;

exports[`WasmBuilder-Utils: ieee754-negative-ints 1`] = `
Array [
  "0,0,0,0",
  "0,0,122,195",
  "0,0,250,195",
  "0,128,59,196",
  "0,0,122,196",
]
`;

exports[`WasmBuilder-Utils: ieee754-positive-float 1`] = `
Array [
  "0,0,0,0",
  "0,0,128,62",
  "0,0,0,63",
  "0,0,64,63",
  "0,0,128,63",
]
`;

exports[`WasmBuilder-Utils: ieee754-positive-ints 1`] = `
Array [
  "0,0,0,0",
  "0,0,122,67",
  "0,0,250,67",
  "0,128,59,68",
  "0,0,122,68",
]
`;

exports[`WasmBuilder-Utils: signedLEB128-negative-ints 1`] = `
Array [
  "0",
  "134,126",
  "140,124",
  "146,122",
  "152,120",
]
`;

exports[`WasmBuilder-Utils: signedLEB128-positive-ints 1`] = `
Array [
  "0",
  "228,0",
  "200,1",
  "172,2",
  "144,3",
  "244,3",
  "216,4",
  "188,5",
  "160,6",
  "132,7",
  "232,7",
]
`;

exports[`WasmBuilder-Utils: unsignedLEB128-positive-ints 1`] = `
Array [
  "0",
  "100",
  "200,1",
  "172,2",
  "144,3",
  "244,3",
  "216,4",
  "188,5",
  "160,6",
  "132,7",
  "232,7",
]
`;

exports[`WasmBuilder-Utils: unsignedLEB128-zero 1`] = `"0"`;
